rusoto/credential/src/lib.rs

```diff
@@ -314,62 +350,69 @@ impl DefaultCredentialsProviderSync {
 /// 3. IAM instance profile. Will only work if running on an EC2 instance with an instance profile/role.
 ///
 /// If the sources are exhausted without finding credentials, an error is returned.
-#[derive(Debug, Default, Clone)]
+#[derive(Debug, Clone)]
 pub struct ChainProvider {
+    instance_metadata_provider: InstanceMetadataProvider,
+    container_provider: ContainerProvider,
     profile_provider: Option<ProfileProvider>,
 }
 
-impl ProvideAwsCredentials for ChainProvider {
-    fn credentials(&self) -> Result<AwsCredentials, CredentialsError> {
+/// Future returned from `ChainProvider`.
+pub struct ChainProviderFuture {
+    inner: Box<Future<Item=AwsCredentials, Error=CredentialsError>>
+}
 
-        EnvironmentProvider
-            .credentials()
-            .or_else(|_| match self.profile_provider {
-                Some(ref provider) => provider.credentials(),
-                None => Err(CredentialsError::new("")),
-            })
-            .or_else(|_| ContainerProvider.credentials())
-            .or_else(|_| InstanceMetadataProvider.credentials())
-            .or_else(|_| {
-                Err(CredentialsError::new(
-                    "Couldn't find AWS credentials in environment, credentials file, or IAM role.",
-                ))
-            })
+impl Future for ChainProviderFuture {
+    type Item = AwsCredentials;
+    type Error = CredentialsError;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        self.inner.poll()
     }
 }
 
-impl ProvideTimeoutableAwsCredentials for ChainProvider {
-    fn credentials_with_timeout(
-        &self,
-        timeout: StdDuration,
-    ) -> Result<AwsCredentials, CredentialsError> {
-        EnvironmentProvider
-            .credentials()
-            .or_else(|_| match self.profile_provider {
-                Some(ref provider) => provider.credentials(),
-                None => Err(CredentialsError::new("")),
-            })
-            .or_else(|_| ContainerProvider.credentials_with_timeout(timeout))
-            .or_else(|_| {
-                InstanceMetadataProvider.credentials_with_timeout(timeout)
+impl ProvideAwsCredentials for ChainProvider {
+    type Future = ChainProviderFuture;
+
+    fn credentials(&self) -> Self::Future {
+        let profile_provider = self.profile_provider.clone();
+        let instance_metadata_provider = self.instance_metadata_provider.clone();
+        let container_provider = self.container_provider.clone();
+        let future = EnvironmentProvider.credentials()
+            .or_else(move |_| match profile_provider {
```

Comment:
The `or_else` chain is how we do the prioritized list of credential sourcing in this PR, right?

Reply:
This is existing behavior, and yes this is how we choose which sources to check first.